network/tcp.go

```diff
@@ -0,0 +1,422 @@
+
+		// Sending 'length' bytes
+		log.Lvl4("Sending from", c.conn.LocalAddr(), "to", c.conn.RemoteAddr())
+		n, err := c.conn.Write(b[:length])
+		if err != nil {
+			log.Error("Couldn't write chunk starting at", sent, "size", length, err)
+			return handleError(err)
+		}
+		sent += Size(n)
+		log.Lvl5("Sent", sent, "out of", packetSize)
+
+		// bytes left to send
+		b = b[n:]
+	}
+	// update stats on the connection
+	c.updateTx(uint64(packetSize))
+	return nil
+}
+
+// Remote returns the name of the peer at the end point of
+// the connection
+func (c *TCPConn) Remote() Address {
+	return c.endpoint
+}
+
+// Local returns the local address and port
+func (c *TCPConn) Local() Address {
+	return NewTCPAddress(c.conn.LocalAddr().String())
+}
+
+// Type implements the Conn interface
+func (c *TCPConn) Type() ConnType {
+	return PlainTCP
+}
+
+// Close ... closes the connection
+func (c *TCPConn) Close() error {
+	c.closedMut.Lock()
+	defer c.closedMut.Unlock()
+	if c.closed == true {
+		return nil
+	}
+	err := c.conn.Close()
+	c.closed = true
+	if err != nil {
+		return handleError(err)
+	}
+	return nil
+}
+
+// handleError produces the higher layer error depending on the type
+// so user of the package can know what is the cause of the problem
+func handleError(err error) error {
+
+	if strings.Contains(err.Error(), "use of closed") || strings.Contains(err.Error(), "broken pipe") {
+		return ErrClosed
+	} else if strings.Contains(err.Error(), "canceled") {
+		return ErrCanceled
+	} else if err == io.EOF || strings.Contains(err.Error(), "EOF") {
+		return ErrEOF
+	}
+
+	netErr, ok := err.(net.Error)
+	if !ok {
+		return ErrUnknown
+	}
+	if netErr.Temporary() {
+		return ErrTemp
+	} else if netErr.Timeout() {
+		return ErrTimeout
+	}
+	return ErrUnknown
+}
+
+// TCPListener is the underlying implementation of
+// Host using Tcp as a communication channel
+type TCPListener struct {
+	// the underlying golang/net listener
+	listener net.Listener
+	// the close channel used to indicate to the listener we want to quit
+	quit chan bool
+	// quitListener is a channel to indicate to the closing function that the
+	// listener has actually really quit
+	quitListener  chan bool
+	listeningLock sync.Mutex
+	listening     bool
+
+	// closed is used to tell the listen routine to return immediatly if a
+	// Stop() have been made (concurrency into place here)
+	closed bool
+
+	// listening addr (actual). Useful for listening on :0 port
+	addr net.Addr
+}
+
+// NewTCPListener returns a Listener. This function tries to bind to the given
+// address already.It returns the listener and an error if one occured during
+// the binding. A subsequent call to Address() will give the real listening
+// address (useful if you set it to port :0).
+func NewTCPListener(addr Address) (*TCPListener, error) {
+	l := &TCPListener{
+		quit:         make(chan bool),
+		quitListener: make(chan bool),
+	}
+	return l, l.bind(addr)
+}
+
+func (t *TCPListener) bind(addr Address) error {
```

Comment:
Why is this separated from `NewTCPListener`? I would merge it with the above. Except if it is called somewhere else.